### ISSUE TITLE
Crash when disk full

### ISSUE THREAD

**Issue by teto** [neovim/neovim#3364]
Hi,

I got a crash when opening a file with my hard drive full. I freed some space and neovim was working again:

```
(gdb) file nvim
Reading symbols from nvim...done.
Reading in symbols for /home/teto/neovim/src/nvim/main.c...done.
(gdb) run resolv/gethnamaddr.c
Starting program: /usr/local/bin/nvim resolv/gethnamaddr.c
Reading symbols from /lib64/ld-linux-x86-64.so.2...Reading symbols from /usr/lib/debug//lib/x86_64-linux-gnu/ld-2.21.so...done.
done.
Reading symbols from system-supplied DSO at 0x7ffff7ffa000...(no debugging symbols found)...done.
Reading in symbols for rtld.c...done.
Reading symbols from /lib/x86_64-linux-gnu/librt.so.1...Reading symbols from /usr/lib/debug//lib/x86_64-linux-gnu/librt-2.21.so...done.
done.
Reading symbols from /lib/x86_64-linux-gnu/libpthread.so.0...Reading symbols from /usr/lib/debug//lib/x86_64-linux-gnu/libpthread-2.21.so...done.
done.
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
Reading symbols from /lib/x86_64-linux-gnu/libdl.so.2...Reading symbols from /usr/lib/debug//lib/x86_64-linux-gnu/libdl-2.21.so...done.
done.
Reading symbols from /lib/x86_64-linux-gnu/libm.so.6...Reading symbols from /usr/lib/debug//lib/x86_64-linux-gnu/libm-2.21.so...done.
done.
Reading symbols from /lib/x86_64-linux-gnu/libutil.so.1...Reading symbols from /usr/lib/debug//lib/x86_64-linux-gnu/libutil-2.21.so...done.
done.
Reading symbols from /lib/x86_64-linux-gnu/libc.so.6...Reading symbols from /usr/lib/debug//lib/x86_64-linux-gnu/libc-2.21.so...done.
done.
Reading in symbols for ../sysdeps/x86_64/strlen.S...done.

Program received signal SIGSEGV, Segmentation fault.
strlen () at ../sysdeps/x86_64/strlen.S:106
106	../sysdeps/x86_64/strlen.S: Aucun fichier ou dossier de ce type.
Current language:  auto
The current source language is "auto; currently asm".
(gdb) bt
Reading in symbols for /home/teto/neovim/src/nvim/memory.c...done.
Reading in symbols for /home/teto/neovim/src/nvim/event/socket.c...done.
Reading in symbols for /home/teto/neovim/src/nvim/msgpack_rpc/server.c...done.
#0  strlen () at ../sysdeps/x86_64/strlen.S:106
#1  0x00000000005c80ea in xstrlcpy (dst=0x7ffff644e580 "\f$\255\373\"]\"%", src=0x0, size=256)
    at /home/teto/neovim/src/nvim/memory.c:362
#2  0x000000000061d785 in socket_watcher_init (loop=0x922940 <loop>, watcher=0x7ffff644e580, endpoint=0x0, data=0x0)
    at /home/teto/neovim/src/nvim/event/socket.c:28
#3  0x000000000061139c in server_start (endpoint=0x0) at /home/teto/neovim/src/nvim/msgpack_rpc/server.c:85
#4  0x000000000061127a in server_init () at /home/teto/neovim/src/nvim/msgpack_rpc/server.c:41
#5  0x00000000004e90a7 in event_init () at /home/teto/neovim/src/nvim/main.c:154
#6  0x00000000004e9254 in main (Reading in symbols for libc-start.c...done.
```
with a very recent version compiled myself
```
NVIM 0.0.0-alpha+201509190915 (compiled Sep 19 2015 16:22:46)
Commit: 2417aeebbffe58eab6ac39feac0f64a27aafbee2
Build type: Debug
Compilation: /usr/bin/cc -Wconversion -g -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -fstack-protector-strong -fdiagnostics-color=auto -DINCLUDE_GENERATED_DECLARATIONS -DHAVE_CONFIG_H -I/home/teto/neovim/build/config -I/home/teto/neovim/src -I/home/teto/neovim/.deps/usr/include -I/home/teto/neovim/.deps/usr/include -I/home/teto/neovim/.deps/usr/include/luajit-2.0 -I/home/teto/neovim/.deps/usr/include -I/home/teto/neovim/.deps/usr/include -I/home/teto/neovim/.deps/usr/include -I/home/teto/neovim/.deps/usr/include -I/usr/include -I/home/teto/neovim/build/src/nvim/auto -I/home/teto/neovim/build/include
Compilé par teto@sirius

Optional features included (+) or not (-): +acl   +iconv    +jemalloc 
For differences from Vim, see :help vim-differences

   fichier vimrc système : "$VIM/nvimrc"
     fichier vimrc utilisateur : "~/.nvimrc"
 2me fichier vimrc utilisateur : "~/.nvim/nvimrc"
      fichier exrc utilisateur : "~/.exrc"
  $VIM par défaut : "/usr/local/share/nvim"
```

**Comment by tarruda** [comment]
This is because `vim_tempname` return nulls if it fails to create a temporary directory(which should happen for a full disk). @teto does this patch fixes the problem?

```diff
diff --git a/src/nvim/msgpack_rpc/server.c b/src/nvim/msgpack_rpc/server.c
index 8dffea3..5b9b6f2 100644
--- a/src/nvim/msgpack_rpc/server.c
+++ b/src/nvim/msgpack_rpc/server.c
@@ -38,6 +38,10 @@ bool server_init(void)
     listen_address = (char *)vim_tempname();
   }

+  if (!listen_address) {
+    return false;
+  }
+
   bool ok = (server_start(listen_address) == 0);
   if (must_free) {
     xfree((char *) listen_address);
```

I think we should also avoid creating a socket by default(no longer necessary after `serverstart()` was implemented by @splinterofchaos

**Comment by teto** [comment]
@tarruda sry I freed some space hence I am not able to reproduce it.
If you want to add it as a test; I saw that method http://bittstriim.blogspot.fr/2009/11/simulating-full-disk-in-linux.html but I am not sure how easy it is to put neovim on such a disk (with all the runtime dependancies etc...).

**Comment by ggVGc** [comment]
Personally, I think "No available disk space" is an exceptional situation, and it's fine for programs to not run at all or correctly in this instance. Having no disk space available destabalizes the whole OS in many ways, hence it's not really useful to spend extra energy on having a program like neovim work cleanly within that scenario. Most programs I know crash and burn in some way or another when there's zero storage available.

**Comment by teto** [comment]
if all programs crash when your disk is full, how do you make space ?
You want to open files

This is not such an exceptional situation, with SSDs, you have small disks,
sometimes you download a huge file in the wrong directory and *BOUM* you
get disk full.
Having vim crash in such situations when you have 12GB ram is kind of sad
(note that vim does not crash with disk full).

2015-09-29 11:04 GMT+02:00 ggVGc <notifications@github.com>:

**Comment by ZyX-I** [comment]
@ggVGc There is big difference regarding *where* “no disk space” is available. If it is /home, /var or /tmp it is still *not* OK to crash, but it is OK to not be fully functional. Having no space on normally read-only / should be absolutely OK, as well as on different other media.

And note: I had such situations with /home and many programs normally were not crashing. I even had or worse: with some bug in who knows where trying to read from/write to /home led to IO error due to crashed XFS driver and this still crashed not so many programs; specifically browsers were badly misbehaving (they have caches there (but old Opera still worked, bug was fixed before they decided to throw away Presto and I thus decided to throw away Opera)) and shells were hanging when you try to open them.

Also note 2: there are file managers for Vim, including built-in netrw. Consider user using them: is it *really* OK to leave user without his favourite file manager when there is no free space? NeoVim may now serve as terminal multiplexer and thus crashing it is even worse idea for *any* reason because most reasons you can imagine are normally solved from shells.

**Comment by jszakmeister** [comment]
It's probably worth noting that hitting quotas can put you in a situation similar to this too.  The OS isn't destabilized in this situation--only the user is restricted from exceeding the quota.

**Comment by BenediktSeidl** [comment]
## TL;DR

I could reproduce the reported issue (the problem seems to be a full /home), the patch fixes it, master is fine without patch

## version as reported

```
NVIM 0.0.0-alpha+201509190915 (compiled Jul 17 2016 08:02:16)
Commit: 2417aeebbffe58eab6ac39feac0f64a27aafbee2
Build type: Debug
Compilation: /usr/bin/cc -Wconversion -g -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -fstack-protector-strong -fdiagnostics-color=auto -DINCLUDE_GENERATED_DECLARATIONS -DHAVE_CONFIG_H -I/home/debian/neovim/build/config -I/home/debian/neovim/src -I/home/debian/neovim/.deps/usr/include -I/home/debian/neovim/.deps/usr/include -I/home/debian/neovim/.deps/usr/include/luajit-2.0 -I/home/debian/neovim/.deps/usr/include -I/home/debian/neovim/.deps/usr/include -I/home/debian/neovim/.deps/usr/include -I/home/debian/neovim/.deps/usr/include -I/usr/include -I/home/debian/neovim/build/src/nvim/auto -I/home/debian/neovim/build/include
Compiled by debian@debian

Optional features included (+) or not (-): +acl   +iconv    +jemalloc 
For differences from Vim, see :help vim-differences

   system vimrc file: "$VIM/nvimrc"
     user vimrc file: "~/.nvimrc"
 2nd user vimrc file: "~/.nvim/nvimrc"
      user exrc file: "~/.exrc"
  fall-back for $VIM: "/usr/local/share/nvim"
```


### full /tmp/, file in /tmp/

#### open

seems to work, error on startup: `E297 Write error in swap file`

#### write (no change):

```
"asd" E514: write error (file system full?)
WARNING: Original file may be lost or damaged
don't quit the editor until the file is successfully written!
Press ENTER or type command to continue
```
file contents are lost after quit

### full /tmp/, full /home/, file in /tmp/

#### open

```
(gdb) run asd
Starting program: /home/debian/neovim/build/bin/nvim asd
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".

Program received signal SIGSEGV, Segmentation fault.
strlen () at ../sysdeps/x86_64/strlen.S:106
106     ../sysdeps/x86_64/strlen.S: No such file or directory.
(gdb) bt
#0  strlen () at ../sysdeps/x86_64/strlen.S:106
#1  0x00000000004d3725 in xstrlcpy (dst=0x7ffff6452300 "", src=0x0, size=256) at /home/debian/neovim/src/nvim/memory.c:362
#2  0x000000000061caa8 in socket_watcher_init (loop=0x91ee80 <loop>, watcher=0x7ffff6452300, endpoint=0x0, data=0x0)
    at /home/debian/neovim/src/nvim/event/socket.c:28
#3  0x0000000000615129 in server_start (endpoint=0x0) at /home/debian/neovim/src/nvim/msgpack_rpc/server.c:85
#4  0x0000000000615007 in server_init () at /home/debian/neovim/src/nvim/msgpack_rpc/server.c:41
#5  0x00000000005c8bf5 in event_init () at /home/debian/neovim/src/nvim/main.c:154
#6  0x00000000005c8da2 in main (argc=2, argv=0x7fffffffe728) at /home/debian/neovim/src/nvim/main.c:249
```

#### version

even `nvim --version` results in the same trace

## version as reported but with patch applied

```
NVIM 0.0.0-alpha+201509190915 (compiled Jul 17 2016 08:30:46)
Commit: 2417aeebbffe58eab6ac39feac0f64a27aafbee2
Build type: Debug
Compilation: /usr/bin/cc -Wconversion -g -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -fstack-protector-strong -fdiagnostics-color=auto -DINCLUDE_GENERATED_DECLARATIONS -DHAVE_CONFIG_H -I/home/debian/neovim/build/config -I/home/debian/neovim/src -I/home/debian/neovim/.deps/usr/include -I/home/debian/neovim/.deps/usr/include -I/home/debian/neovim/.deps/usr/include/luajit-2.0 -I/home/debian/neovim/.deps/usr/include -I/home/debian/neovim/.deps/usr/include -I/home/debian/neovim/.deps/usr/include -I/home/debian/neovim/.deps/usr/include -I/usr/include -I/home/debian/neovim/build/src/nvim/auto -I/home/debian/neovim/build/include
Compiled by debian@debian

Optional features included (+) or not (-): +acl   +iconv    +jemalloc 
For differences from Vim, see :help vim-differences

   system vimrc file: "$VIM/nvimrc"
     user vimrc file: "~/.nvimrc"
 2nd user vimrc file: "~/.nvim/nvimrc"
      user exrc file: "~/.exrc"
  fall-back for $VIM: "/usr/local/share/nvim"
```

### full /tmp and /home, file in /tmp

#### opening file

```
E483: Can't get temp file name
E483: Can't get temp file name
Press ENTER or type command to continue
```
then `E297 Write error in swap file` but works

#### writing (no change)

same error as "writing" above, file content lost

## version: master

```
NVIM v0.1.5-627-g7fa1baf
Build type: Debug
Compilation: /usr/bin/cc -Wconversion -g -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wvla -fstack-protector-strong -fdiagnostics-color=auto -DINCLUDE_GENERATED_DECLARATIONS -DHAVE_CONFIG_H -D_GNU_SOURCE -I/home/debian/neovim/build/config -I/home/debian/neovim/src -I/home/debian/neovim/.deps/usr/include -I/home/debian/neovim/.deps/usr/include -I/home/debian/neovim/.deps/usr/include -I/home/debian/neovim/.deps/usr/include -I/home/debian/neovim/.deps/usr/include -I/home/debian/neovim/.deps/usr/include -I/usr/include -I/home/debian/neovim/build/src/nvim/auto -I/home/debian/neovim/build/include
Compiled by debian@debian

Optional features included (+) or not (-): +acl   +iconv    +jemalloc +tui      
For differences from Vim, see :help vim-differences

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/usr/local/share/nvim"
```

### full /tmp and /home, file in /tmp

#### opening file

```
E484: Can't open file /usr/local/share/nvim/syntax/syntax.vim
Press ENTER or type command to continue
```
then
```
E303: Unable to create directory "/home/debian/.local" for swap file, recovery impossible: no space left on device
"asd" 1L, 11C
Press ENTER or type command to continue
```

but opens, additional error on quit:

```
E886: Failed to create directory /home/debian/.local for writing ShaDa file: no space left on device
Press ENTER or type command to continue
```


#### writing file

same as "writing" above, file content lost

### full /tmp and /home and /var and /, file in /tmp

#### opening file

same as "full /tmp and /home"

**Comment by justinmk** [comment]
@BenediktSeidl Thanks for investigating. It looks like the latest version doesn't reach the segfaulting codepath, but we should still apply the patch.
